src/icelake/icelake_utf8_length_from_utf16.inl.cpp

```diff
@@ -1,84 +1,292 @@
+
+      __m512i lb_masked2 =
+          _mm512_and_si512(current2, _mm512_set1_epi16(uint16_t(0xfc00)));
+      __m512i block_masked2 =
+          _mm512_and_si512(next2, _mm512_set1_epi16(uint16_t(0xfc00)));
+      __mmask32 lb_is_high2 = _mm512_cmpeq_epi16_mask(
+          lb_masked2, _mm512_set1_epi16(uint16_t(0xd800)));
+      __mmask32 block_is_low2 = _mm512_cmpeq_epi16_mask(
+          block_masked2, _mm512_set1_epi16(uint16_t(0xdc00)));
+      __mmask32 illseq2 = _kxor_mask32(lb_is_high2, block_is_low2);
+      mismatched_count += count_ones32(illseq1);
+      mismatched_count += count_ones32(illseq2);
+    }
+
+    count += count_ones32(c01);
+    count += count_ones32(c11);
+    count -= count_ones32(is_surrogate1);
+    count += count_ones32(c02);
+    count += count_ones32(c12);
+    count -= count_ones32(is_surrogate2);
   }
+  if (pos + N + 1 <= size) {
+    __m512i input =
+        _mm512_loadu_si512(reinterpret_cast<const __m512i *>(in + pos));
+    if (!match_system(big_endian)) {
+      input = _mm512_shuffle_epi8(input, byteflip);
+    }
+    __m512i input_next =
+        _mm512_loadu_si512(reinterpret_cast<const __m512i *>(in + pos + 1));
+    if (!match_system(big_endian)) {
+      input_next = _mm512_shuffle_epi8(input_next, byteflip);
+    }
+
+    __m512i lb_masked =
+        _mm512_and_si512(input, _mm512_set1_epi16(uint16_t(0xfc00)));
+    __m512i block_masked =
+        _mm512_and_si512(input_next, _mm512_set1_epi16(uint16_t(0xfc00)));
+    __mmask32 lb_is_high =
+        _mm512_cmpeq_epi16_mask(lb_masked, _mm512_set1_epi16(uint16_t(0xd800)));
+    __mmask32 block_is_low = _mm512_cmpeq_epi16_mask(
+        block_masked, _mm512_set1_epi16(uint16_t(0xdc00)));
+    __mmask32 illseq = _kxor_mask32(lb_is_high, block_is_low);
+    __mmask32 is_surrogate = _mm512_cmpeq_epi16_mask(
+        _mm512_and_si512(input, _mm512_set1_epi16(uint16_t(0xf800))),
+        _mm512_set1_epi16(uint16_t(0xd800)));
+    __mmask32 c0 =
+        _mm512_test_epi16_mask(input, _mm512_set1_epi16(uint16_t(0xff80)));
+    __mmask32 c1 =
+        _mm512_test_epi16_mask(input, _mm512_set1_epi16(uint16_t(0xf800)));
 
-  if (iteration > 0) {
-    count += v_count.sum();
+    count += count_ones32(c0);
+    count += count_ones32(c1);
+    count -= count_ones32(is_surrogate);
+    mismatched_count += count_ones32(illseq);
+    pos += N;
   }
 
-  return count + scalar::utf16::utf8_length_from_utf16<big_endian>(in + pos,
-                                                                   size - pos);
+  __mmask32 remaining_mask =
+      0xFFFFFFFFULL >>
+      (32 - (size - pos)); // mask for the remaining char16 values
+  // for remaining_next_mask, we start from 0x7FFFFFFFULL instead of
+  // 0xFFFFFFFFULL We could also just shift right by one remaining_mask.
+  __mmask32 remaining_next_mask = 0x7FFFFFFFULL >> (32 - (size - pos));
+  __m512i input = _mm512_maskz_loadu_epi16(remaining_mask, in + pos);
```

Comment:
I don't understand how this load doesn't segfault on strings near page edges. Perhaps I'm overlooking something, but I think you should step back by the number of positions the mask was shifted by and shift the mask the other way. Since we are in the large-input function we can't segfault at the beginning by doing this.